functorch/experimental/map.py

```diff
@@ -0,0 +1,97 @@
+import torch
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard
+from torch._ops import PyOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    get_proxy_slot,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch.utils._pytree import tree_flatten
+
+
+map = PyOperator("map")
+
+
+def trace_map(proxy_mode, func_overload, f, xs, *args):
+    def _unwrap_proxy(e):
+        if not isinstance(e, (torch.Tensor, torch.SymInt, torch.SymFloat)):
+            return e
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(xs, torch.Tensor), "map() must loop over a tensor"
+    assert len(xs.shape) > 0 and xs.shape[0] > 0, "map() cannot be traced with scalar tensors or zero dimension tensors"
+    assert all(isinstance(o, (torch.Tensor, torch.nn.Module)) for o in args), "map() operands must be a list of tensors of modules"
```

Comment:
nit: should be RuntimeError instead of assertion, these are user-facing errors, right?

Reply:
sure